### PR TITLE
Simple array functions

### PR DESCRIPTION
**Pull request by bschommer**
This PR adds the functions `Array.exists` and `Array.for_all` to the standard library.  The new functions mirror their counterparts in the `List` module of the standard library.

As needed there are a few tests for the new functions in the testsuite/tests/array-functions folder and as any other developer can recreate the changes without access to the original patch would write an indistinguishable patch I think a CLA is not needed.

### CODE REVIEW

**Comment by Gbury** [comment]
I think the functions should return as soon as possible. Currently, they always traverse the whole array, which we should definitely avoid whenever possible.

**Comment by bschommer** [comment]
You are right. I updated the functions to exit earlier.

**Comment by alainfrisch** [comment]
Thanks!

What about `Array.exists2`, `Array.for_all2`, for the sake of uniformity with List?

**Comment by bschommer** [comment]
I added `Array.exist2` and `Array.for_all2` and also `Array.mem`, `Array.memq` and `Array.find`

File: testsuite/tests/array-functions/test.ml [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  assert (Array.memq 2 a);
+  assert (Array.memq 3 a);
+  assert (Array.memq 4 a);
+  assert (Array.memq 5 a);
+  assert (Array.memq 6 a);
+  assert (Array.memq 7 a);
+  assert (Array.memq 8 a);
+  assert (Array.memq 9 a);
+  assert (not (Array.memq 0 a));
+  assert (not (Array.memq 10 a));
```
Comment (gasche) [inline testsuite/tests/array-functions/test.ml:167]: `mem` and `memq` need some testing on non-immediate values (arrays, references, lists, blocks in general) to exercise non-trivial identity. See the [Batteries tests](https://github.com/ocaml-batteries-team/batteries-included/blob/master/src/batArray.mlv#L142-L161), for example.

File: stdlib/array.ml [hunk 2]
```diff
@@ -134,6 +134,66 @@ let fold_right f a x =
...
+  else let rec loop i =
+    if i = n1 then true
+    else if p (unsafe_get l1 i) (unsafe_get l2 i) then loop (succ i)
+    else false in
+  loop 0
+
+let exists2 p l1 l2 =
+  let n1 = length l1
+  and n2 = length l2 in
+  if n1 <> n2 then invalid_arg "Array.exists2"
```
Comment (yallop) [inline stdlib/array.ml]: For consistency with `List.exists2`, the function should only raise `Invalid_argument` if the search reaches the end of one of the arrays.  For example, the following should return `true`:

`Array.exists2 (=) [|1; 2|] [|1|]`

File: stdlib/array.ml [hunk 3]
```diff
@@ -134,6 +134,66 @@ let fold_right f a x =
...
+    if i = n1 then false
+    else if p (unsafe_get l1 i) (unsafe_get l2 i) then true
+    else loop (succ i) in
+  loop 0
+
+let mem x a =
+  let n = length a in
+  let rec loop i =
+    if i = n then false
+    else if compare (unsafe_get a i) x == 0 then true
```
Comment (yallop) [inline stdlib/array.ml]: This should be `=`, not `==`.

File: Changes [hunk 4]
```diff
@@ -197,6 +197,9 @@ Standard library:
   * Pervasives.ldexp (and [@@noalloc])
   * Pervasives.compare for float, nativeint, int32, int64.
   (Bobot François)
+- GPR#329: Add exist, for_all, exist2,  for_all2, mem, memq and find
```
Comment (yallop) [inline Changes]: `exist2` ~> `exists2`

File: stdlib/array.mli [hunk 5]
```diff
@@ -152,6 +152,40 @@ val fold_right : ('b -> 'a -> 'a) -> 'b array -> 'a -> 'a
...
+   to have different lengths. *)
+
+val exists2 : ('a -> 'b -> bool) -> 'a array -> 'b array -> bool
+(** Same as {!Array.exists}, but for a two-argument predicate.
+   Raise [Invalid_argument] if the two arrays are determined
+   to have different lengths. *)
+
+val mem : 'a -> 'a array -> bool
+(** [mem a l] is true if and only if [a] is equal
+   to an element of [l]. *)
```
Comment (yallop) [inline stdlib/array.mli:167]: It'd be better to use `[a]` instead of `[l]` for the array argument (here and throughout).

**Comment by gasche** [comment]
I'm not convinced by the current way the testsuite is written for those functions. Having arrays of size 9 with 9 different tests does not bring us noticeably more safety than doing the same thing for arrays of size, say, 3 (the only case I see where this would matter is a complex partially-unfolded implementation that handles N elements at once for N above 3 and below 9, but I would expect the proposer of such a complex implementation to improve the testsuite accordingly anyway). On the other hand, the corner cases of the empty array and a float array are not tested and should be. (Then for all functions performing comparisons on the elements we need tests with element types with non-trivial comparison, such as integer references)

**Comment by bschommer** [comment]
I will add more tests also for the corner cases of array of floats and arrays of refs etc. for mem and memq. I was not sure how much tests I needed to add.

**Comment by bschommer** [comment]
I added even more tests. Now we have more test with refs and also arrays of arrays.

**Comment by gasche** [comment]
Thinking more about it, I'm not convinced that adding `exists2` and `for_all2` is a good idea. @yallop pointed a subtlety of whether to fail eagerly or lazily when the two arrays are of different sizes. Jeremy recommends the lazy approach (consistent with `List` functions where size is not known in advance), but Batteries function of the same name and type use the eager approach, and I am not very excited at the idea of a difference in behavior between both.

I think we are going out of the "obvious functions only" mandate that is the compromise for 4.03.

**Comment by gasche** [comment]
I would be in favor of merging a pull request with `exists`, `for_all`, `mem` and `memq` that are non-controversial. I'm against merging `for_all2` and `exists2` for the reason above, and I have vague worries about `find` that I think would be best avoided for now.

**Comment by bschommer** [comment]
I understand your concern and actually my first motivation for adding functions was to add a `find` function that also returns the index but a quick scan of existing implementations convinced my otherwise.
From my side there is no problem in removing `for_all2` and `exists2`, also `find`.

**Comment by dbuenzli** [comment]
> I have vague worries about find that I think would be best avoided for now.

Please do not perpetuate design errors for the sake of consistency (i.e. `Not_found` exceptions).

Also I would personally be interested in a solid stdlib improvement release rather that these things going in for 4.03. Adding a little function here and there on each OCaml release is very annoying for everyone involved: if you use the function you suddenly end up depending on a major release of OCaml for no real good reason and people ask you not to that or you discover it only once you submit to the opam repo. This means that in the end you still have to rewrite the thing to support earlier versions and a lot of time is being wasted on many ends. This way of proceeding also tend to fail to see the greater design.

**Comment by alainfrisch** [comment]
> From my side there is no problem in removing for_all2 and exists2, also find.

Ok, let's do that.  Apologies to @bschommer mer for asking you to go back and forth.

> Adding a little function here and there on each OCaml release is very annoying for everyone involved

I hear you, but I'm opposed to rejecting such changes.  For me, it remains to be proved that the community can manage to produce "the next stdlib", and I would hate if waiting for it to happen blocked any incremental improvement.  Not everybody need or insist on maintaining compatibility with older releases.

The arguments you provide would apply equally well to any change to the language.  

Perhaps we could have proper `[@@ocaml.since]` attributes to mark the version in which a specific function has been introduced, so as to support automated check that a given project doesn't use overly recent stdlib or language features.

**Comment by yallop** [comment]
> Adding a little function here and there on each OCaml release is very annoying for everyone involved: if you use the function you suddenly end up depending on a major release of OCaml for no real good reason and people ask you not to that or you discover it only once you submit to the opam repo. This means that in the end you still have to rewrite the thing to support earlier versions and a lot of time is being wasted on many ends.

This is a good point, and we've seen several such problems on opam-repository, particularly with `|>`.

One way to mitigate the problem would be to have library compatibility packages along the lines of [base-bytes](https://github.com/ocaml/opam-repository/tree/master/packages/base-bytes), to make it possible to use recent standard library additions with older OCaml releases.  I think having one package per library version (e.g. a `stdlib-4.02.3` package, available for all previous OCaml versions) would be easier to manage than (e.g.) one package per feature.

**Comment by dbuenzli** [comment]
> For me, it remains to be proved that the community can manage to produce "the next stdlib",

I'm not necessarily talking about the "next stdlib" but the let's resolve the little inconsistencies like these ones stdlib.

> The arguments you provide would apply equally well to any change to the language.

I personally wouldn't equate language changes with addition of a function in the stdlib.

**Comment by bschommer** [comment]
As suggested I removed for_all2, exists2 and find.

**Comment by yallop** [comment]
@gasche:
> Thinking more about it, I'm not convinced that adding `exists2` and `for_all2` is a good idea. @yallop pointed a subtlety of whether to fail eagerly or lazily when the two arrays are of different sizes. Jeremy recommends the lazy approach (consistent with `List` functions where size is not known in advance), but Batteries function of the same name and type use the eager approach, and I am not very excited at the idea of a difference in behavior between both.

For what it's worth, I've changed my mind on this: I no longer have any preference between the two approaches to handling mismatched array sizes.  However, I would quite like to have `exists2` and `for_all2` in the `Array` library.  Is the backwards-compatibility argument against their inclusion still so strong?  There have been quite a few small additions to the standard library since this discussion, so perhaps preferences have changed.

**Comment by gasche** [comment]
I think you should feel free to submit a separate PR (authorship could even be credited to @bschommer if you reuse the code) to suggest `exists2` and `for_all2`. I would recommend the strict behavior, but I also don't feel strongly about it.

**Comment by yallop** [comment]
I've done just that: see #9235.
